src/pybind/mgr/cephadm/module.py

```diff
@@ -1947,54 +1947,42 @@ def get_osd_uuid_map(self, only_up=False):
                 r[str(o['osd'])] = o['uuid']
         return r
 
-    def call_inventory(self, hosts, drive_groups):
-        def call_create(inventory):
-            return self._prepare_deployment(hosts, drive_groups, inventory)
-
-        return self.get_inventory().then(call_create)
-
-    def create_osds(self, drive_groups):
-        # type: (List[DriveGroupSpec]) -> AsyncCompletion
-        return self.get_hosts().then(lambda hosts: self.call_inventory(hosts, drive_groups))
-
-    def _prepare_deployment(self,
-                            all_hosts,  # type: List[orchestrator.HostSpec]
-                            drive_groups,  # type: List[DriveGroupSpec]
-                            inventory_list  # type: List[orchestrator.InventoryHost]
-                            ):
-        # type: (...) -> orchestrator.Completion
-
-        for drive_group in drive_groups:
-            self.log.info("Processing DriveGroup {}".format(drive_group))
-            # 1) use fn_filter to determine matching_hosts
-            matching_hosts = drive_group.placement.pattern_matches_hosts([x.hostname for x in all_hosts])
-            # 2) Map the inventory to the InventoryHost object
-            # FIXME: lazy-load the inventory from a InventoryHost object;
-            #        this would save one call to the inventory(at least externally)
-
-            def _find_inv_for_host(hostname, inventory_list):
-                # This is stupid and needs to be loaded with the host
-                for _inventory in inventory_list:
-                    if _inventory.name == hostname:
-                        return _inventory
-                raise OrchestratorError("No inventory found for host: {}".format(hostname))
-
-            cmds = []
-            # 3) iterate over matching_host and call DriveSelection and to_ceph_volume
-            for host in matching_hosts:
-                inventory_for_host = _find_inv_for_host(host, inventory_list)
-                drive_selection = selector.DriveSelection(drive_group, inventory_for_host.devices)
-                cmd = translate.to_ceph_volume(drive_group, drive_selection).run()
-                if not cmd:
-                    self.log.info("No data_devices, skipping DriveGroup: {}".format(drive_group.service_id))
-                    continue
-                cmds.append((host, cmd))
-
-        return self._create_osds(cmds)
-
-    @async_map_completion
-    def _create_osds(self, host, cmd):
-        return self._create_osd(host, cmd)
+    def apply_drivegroups(self, specs: List[DriveGroupSpec]) -> List[orchestrator.Completion]:
+        completions: List[orchestrator.Completion] = list()
+        for spec in specs:
+            completions.extend(self._apply(spec))
+        return completions
```

Comment:
```suggestion
    def apply_drivegroups(self, specs: List[DriveGroupSpec]) -> orchestrator.Completion:
        completion: orchestrator.Completion = self._apply(specs[0])
        for spec in specs[1:]:
            completion.then(self._apply(spec))
        return completions
```

Reply:
do you really think that this is superior to having a list of completions that will be processed in the orch_cli module? I'd sort of agree if the completions were long running and should fail fast, but at this point it feels minuscule.

Reply:
well,actually I think `_apply()` shoudn't return a completion at all and instead just return a string.

Reply:
i'm wondering if anything should return a completion... since the only thing we ever do with a completion is block on it?

Reply:
as long as we raise on failure (and the exception is being presented to the user) I think we can drop the completions.

Reply:
Let me have a look, if we can drop completions entirely.

Reply:
I'd recommend to make the changes in bulk in a separate PR since we need to streamline all the `apply_$service` methods.

Reply:
I would simplify this by 
- ``apply_drivegroups`` -> ``apply_osd``
- take a single spec, not a list, so that it can just call directly into _apply() like all of the others
...but happy to clean that up later.